webbpsf/wfirst.py

```diff
@@ -323,5 +323,105 @@ def _validateConfig(self, **kwargs):
+            self.filter = filter
+        elif mode is 'CHARSPC':
+            self.filter = 'F770'
+        else:
+            self.filter = 'F721'
+
+    def _validate_config(self):
+        if self.mode in self.mode_list:
+            return True
+
+    def _addAdditionalOptics(self, optsys, oversample=4):
+        """Add coronagraphic or spectrographic optics for WFIRST CGI."""
+    
+        trySAM = False
+        char_fpmres = 4
+        disk_fpmres = 3
+
+        if ('pupil_shift_x' in self.options and self.options['pupil_shift_x'] != 0) or \
+           ('pupil_shift_y' in self.options and self.options['pupil_shift_y'] != 0):
+            shift = (self.options['pupil_shift_x'], self.options['pupil_shift_y'])
+        else: shift = None
+
+        if self.mode == 'CHARSPC':
+            optsys.addPupil(transmission=os.path.join(self._datapath, "optics/CHARSPC_SP_256pix.fits.gz"), name=self.mode, shift=None)
+        else: # self.mode == 'DISKSPC':
+            optsys.addPupil(transmission=os.path.join(self._datapath, "optics/DISKSPC_SP_256pix.fits.gz"), name=self.mode, shift=None)
+        if self.mode == 'CHARSPC':
+            if self.filter == 'F660':
+                self.image_mask = 'CHARSPC_F660'
+                optsys.addImage(transmission=os.path.join(self._datapath,"optics/CHARSPC_FPM_25WA90_2x65deg_-_FP1res{0:d}_evensamp_D{1:03d}_F660.fits.gz".format(char_fpmres, 2*9*char_fpmres)))
+            elif self.filter == 'F770':
+                self.image_mask = 'CHARSPC_F770'
+                optsys.addImage(transmission=os.path.join(self._datapath,"optics/CHARSPC_FPM_25WA90_2x65deg_-_FP1res{0:d}_evensamp_D{1:03d}_F770.fits.gz".format(char_fpmres, 2*9*char_fpmres)))
+            elif self.filter == 'F890':
+                self.image_mask = 'CHARSPC_F890'
+                optsys.addImage(transmission=os.path.join(self._datapath,"optics/CHARSPC_FPM_25WA90_2x65deg_-_FP1res{0:d}_evensamp_D{1:03d}_F890.fits.gz".format(char_fpmres, 2*9*char_fpmres)))
+        else: #elif self.filter == 'F721':
+            self.image_mask = 'DISKSPC_F721'
+            optsys.addImage(transmission=os.path.join(self._datapath,"optics/DISKSPC_FPM_65WA200_360deg_-_FP1res{0:d}_evensamp_D{1:03d}_F721.fits.gz".format(disk_fpmres, 2*20*disk_fpmres)))
+
+        if self.mode == 'CHARSPC':
+            if not hasattr(self, 'fov_arcsec') or not self._override_fov:
+                self.fov_arcsec = 2*0.82 # See 2015 SDT report, Section 3.4.1.1.1: IFS has 76 lenslets across the (2 x 0.82) arcsec FoV.
+            if not hasattr(self, 'pixelscale') or not self._override_pixelscale:
+                self.pixelscale = 0.025 # Nyquist at 600 nm
+        elif self.mode == 'DISKSPC':
+            if not hasattr(self, 'fov_arcsec') or not self._override_fov:
+                self.fov_arcsec = 3.2
+            if not hasattr(self, 'pixelscale') or not self._override_pixelscale:
+                self.pixelscale = 0.020 # Nyquist at 465 nm
+        # Lyot stop
```

Comment:
One way to minimize the opportunities for meaningless configurations is to have some of these attributes implemented as properties which are calculated on the basis of others behind the scenes. See the NIRCam class for examples of this - it presents attributes for `channel` (long or short) and `module` (A or B) which are calculated properties. So the 'camera' attribute might be calculated based on the coronagraph selection, if you think that makes sense. NIRCam also automatically toggles the pixel scale based on the `channel`, so that logic is exactly parallel.